File: nodejs/packages/periph/src/chips/magnetometer/hmc5883l.js
'use strict';

const _REG_CONFIG_A   = 0x00;
const _REG_CONFIG_B   = 0x01;
const _REG_MODE       = 0x02;
const _REG_DATA_X_MSB = 0x03;
const _REG_STATUS     = 0x09;
const _REG_ID_A       = 0x0A;
const _REG_ID_B       = 0x0B;
const _REG_ID_C       = 0x0C;

const _GAIN_LSB_PER_GAUSS = {
    0: 1370,  // GN=0: ±0.88 Ga
    1: 1090,  // GN=1: ±1.3 Ga (default)
    2: 820,   // GN=2: ±1.9 Ga
    3: 660,   // GN=3: ±2.5 Ga
    4: 440,   // GN=4: ±4.0 Ga
    5: 390,   // GN=5: ±4.7 Ga
    6: 330,   // GN=6: ±5.6 Ga
    7: 230,   // GN=7: ±8.1 Ga
};

/**
 * HMC5883L 3-axis magnetometer — minimal interface.
 *
 * Reads magnetic field on all three axes in continuous mode with sensible
 * defaults baked in. No configuration required beyond the connection.
 *
 * Default behaviour (baked into Minimal):
 * - Averaging: 8 samples (MA=11)
 * - ODR: 15 Hz (DO=100)
 * - Gain: ±1.3 Ga (GN=001), 1090 LSb/Gauss
 * - Mode: continuous measurement
 */
class HMC5883LMinimal {
    /**
     * @param {import('../../connection/connection').Connection} connection - Configured I²C connection pointing at the device (fixed address 0x1E).
     */
    constructor(connection) {
        this._conn = connection;
        this._gain = 1;
        this._gainLsbPerGauss = _GAIN_LSB_PER_GAUSS[this._gain];
        this._initMinimal();
    }

    async _initMinimal() {
        await this._writeReg8(_REG_CONFIG_A, 0x70);
        await this._writeReg8(_REG_CONFIG_B, 0x20);
        await this._writeReg8(_REG_MODE, 0x00);
        // 6 ms delay for first measurement - caller should wait if needed
    }

    async _readReg8(reg) {
        return (await this._conn.writeRead(Buffer.from([reg]), 1))[0];
    }

    async _readReg16(reg) {
        const raw = await this._conn.writeRead(Buffer.from([reg]), 2);
        return raw.readInt16BE(0);
    }

    async _writeReg8(reg, value) {
        const buf = Buffer.alloc(2);
        buf[0] = reg;
        buf[1] = value;
        await this._conn.write(buf);
    }

    async _readDataBurst() {
        const raw = await this._conn.writeRead(Buffer.from([_REG_DATA_X_MSB]), 6);
        const rawX = raw.readInt16BE(0);
        const rawZ = raw.readInt16BE(2);
        const rawY = raw.readInt16BE(4);
        return { rawX, rawY, rawZ };
    }

    _rawToTesla(raw) {
        if (raw === -4096) {
            return null;
        }
        return (raw / this._gainLsbPerGauss) * 1e-4;
    }

    /**
     * Read magnetic field on all three axes.
     * @returns {Promise<{x: number|null, y: number|null, z: number|null}>} X, Y, Z field strength in Tesla.
     * Returns null for any axis that overflows (raw == -4096).
     */
    async magneticField() {
        const { rawX, rawY, rawZ } = await this._readDataBurst();
        return {
            x: this._rawToTesla(rawX),
            y: this._rawToTesla(rawY),
            z: this._rawToTesla(rawZ),
        };
    }
}

/**
 * HMC5883L full interface — extends HMC5883LMinimal with complete chip functionality.
 *
 * Adds configuration, single-shot mode, self-test, identification, and status access.
 */
class HMC5883LFull extends HMC5883LMinimal {
    /**
     * @param {import('../../connection/connection').Connection} connection - Configured I²C connection pointing at the device (fixed address 0x1E).
     */
    constructor(connection) {
        super(connection);
    }

    /**
     * Write Configuration Registers A and B.
     * @param {number} [odr=15] - Data output rate in Hz (continuous mode). Valid: 0.75, 1.5, 3, 7.5, 15, 30, 75.
     * @param {number} [averaging=8] - Samples averaged per output. Valid: 1, 2, 4, 8.
     * @param {number} [gain=1] - Gain index 0–7.
     * @returns {Promise<void>}
     */
    async configure(odr = 15, averaging = 8, gain = 1) {
        const maMap = { 1: 0b00, 2: 0b01, 4: 0b10, 8: 0b11 };
        const doMap = { 0.75: 0b000, 1.5: 0b001, 3: 0b010, 7.5: 0b011,
                        15: 0b100, 30: 0b101, 75: 0b110 };

        if (!maMap[averaging]) throw new Error('averaging must be 1, 2, 4, or 8');
        if (!doMap[odr]) throw new Error('odr must be 0.75, 1.5, 3, 7.5, 15, 30, or 75');
        if (gain < 0 || gain > 7) throw new Error('gain must be 0–7');

        const ma = maMap[averaging];
        const doBits = doMap[odr];
        const configA = (ma << 5) | (doBits << 2);
        await this._writeReg8(_REG_CONFIG_A, configA);

        const configB = (gain << 5);
        await this._writeReg8(_REG_CONFIG_B, configB);

        this._gain = gain;
        this._gainLsbPerGauss = _GAIN_LSB_PER_GAUSS[gain];
    }

    /**
     * Update the gain setting (GN bits in Config B).
     * @param {number} gain - Gain index 0–7.
     * @returns {Promise<void>}
     */
    async setGain(gain) {
        if (gain < 0 || gain > 7) throw new Error('gain must be 0–7');
        await this._writeReg8(_REG_CONFIG_B, gain << 5);
        this._gain = gain;
        this._gainLsbPerGauss = _GAIN_LSB_PER_GAUSS[gain];
    }

    /**
     * Set the operating mode.
     * @param {string} mode - 'continuous', 'single', or 'idle'.
     * @returns {Promise<void>}
     */
    async setMode(mode) {
        const modeMap = { continuous: 0b00, single: 0b01, idle: 0b10 };
        if (!modeMap[mode]) throw new Error("mode must be 'continuous', 'single', or 'idle'");
        await this._writeReg8(_REG_MODE, modeMap[mode]);
    }

    /**
     * Check if new measurement data is ready.
     * @returns {Promise<boolean>} True if RDY bit is set in Status Register.
     */
    async dataReady() {
        const status = await this._readReg8(_REG_STATUS);
        return !!(status & 0x01);
    }

    /**
     * Read the raw Status Register.
     * @returns {Promise<number>} Raw STATUS register byte (RDY in bit 0, LOCK in bit 1).
     */
    async status() {
        return this._readReg8(_REG_STATUS);
    }

    /**
     * Take a single measurement in single-shot mode.
     * @returns {Promise<{x: number|null, y: number|null, z: number|null}>} X, Y, Z field strength in Tesla.
     * Returns null for any axis that overflows (raw == -4096).
     */
    async singleMeasurement() {
        await this._writeReg8(_REG_MODE, 0x01);
        // Caller should wait 6 ms before reading
        return this.magneticField();
    }

    /**
     * Read the identification registers.
     * @returns {Promise<[number, number, number]>} (id_a, id_b, id_c) — expected (0x48, 0x34, 0x33) = ASCII 'H43'.
     */
    async identify() {
        const idA = await this._readReg8(_REG_ID_A);
        const idB = await this._readReg8(_REG_ID_B);
        const idC = await this._readReg8(_REG_ID_C);
        return [idA, idB, idC];
    }

    /**
     * Run self-test with positive or negative bias.
     * @param {boolean} [positive=true] - True for positive bias (MS=01), false for negative bias (MS=10).
     * @returns {Promise<{x: number|null, y: number|null, z: number|null}>} X, Y, Z field deflection in Tesla.
     * Returns null for any axis that overflows.
     */
    async selfTest(positive = true) {
        const configA = await this._readReg8(_REG_CONFIG_A);
        const ms = positive ? 0b01 : 0b10;
        await this._writeReg8(_REG_CONFIG_A, (configA & 0xFC) | ms);

        await this._writeReg8(_REG_MODE, 0x01);
        // Caller should wait 6 ms
        const result = await this.magneticField();

        await this._writeReg8(_REG_CONFIG_A, (configA & 0xFC) | 0b00);
        return result;
    }
}

module.exports = { HMC5883LMinimal, HMC5883LFull };

File: cpp/src/chips/magnetometer/HMC5883L.cpp
#include "HMC5883L.h"
#include <cmath>

HMC5883LMinimal::HMC5883LMinimal(Connection& connection)
    : _connection(connection), _gain(1), _gain_lsb_per_gauss(GAIN_LSB_PER_GAUSS[1]) {
    _init_minimal();
}

void HMC5883LMinimal::_init_minimal() {
    _write_reg8(REG_CONFIG_A, 0x70);
    _write_reg8(REG_CONFIG_B, 0x20);
    _write_reg8(REG_MODE, 0x00);
    // 6 ms delay for first measurement - platform-specific delay needed
    // For now, we assume the platform handles timing or user calls after delay
}

uint8_t HMC5883LMinimal::_read_reg8(uint8_t reg) {
    uint8_t buf[1];
    _connection.write_read(&reg, 1, buf, 1);
    return buf[0];
}

int16_t HMC5883LMinimal::_read_reg16(uint8_t reg) {
    uint8_t buf[2];
    _connection.write_read(&reg, 1, buf, 2);
    return (int16_t)(((uint16_t)buf[0] << 8) | buf[1]);
}

void HMC5883LMinimal::_write_reg8(uint8_t reg, uint8_t value) {
    uint8_t buf[2] = { reg, value };
    _connection.write(buf, 2);
}

void HMC5883LMinimal::_read_data_burst(int16_t& raw_x, int16_t& raw_y, int16_t& raw_z) {
    uint8_t buf[6];
    _connection.write_read(&REG_DATA_X_MSB, 1, buf, 6);
    raw_x = (int16_t)(((uint16_t)buf[0] << 8) | buf[1]);
    raw_z = (int16_t)(((uint16_t)buf[2] << 8) | buf[3]);
    raw_y = (int16_t)(((uint16_t)buf[4] << 8) | buf[5]);
}

float HMC5883LMinimal::_raw_to_tesla(int16_t raw) {
    if (raw == -4096) {
        return NAN;
    }
    return (raw / _gain_lsb_per_gauss) * 1e-4f;
}

bool HMC5883LMinimal::magnetic_field(float& x, float& y, float& z) {
    int16_t raw_x, raw_y, raw_z;
    _read_data_burst(raw_x, raw_y, raw_z);
    x = _raw_to_tesla(raw_x);
    y = _raw_to_tesla(raw_y);
    z = _raw_to_tesla(raw_z);
    return !std::isnan(x) && !std::isnan(y) && !std::isnan(z);
}

// HMC5883LFull

HMC5883LFull::HMC5883LFull(Connection& connection)
    : HMC5883LMinimal(connection) {}

bool HMC5883LFull::configure(float odr, uint8_t averaging, uint8_t gain) {
    uint8_t ma, do_bits;

    // Map averaging
    switch (averaging) {
        case 1:  ma = 0b00; break;
        case 2:  ma = 0b01; break;
        case 4:  ma = 0b10; break;
        case 8:  ma = 0b11; break;
        default: return false;
    }

    // Map ODR
    if (fabsf(odr - 0.75f) < 0.01f)        do_bits = 0b000;
    else if (fabsf(odr - 1.5f) < 0.01f)    do_bits = 0b001;
    else if (fabsf(odr - 3.0f) < 0.01f)    do_bits = 0b010;
    else if (fabsf(odr - 7.5f) < 0.01f)    do_bits = 0b011;
    else if (fabsf(odr - 15.0f) < 0.01f)   do_bits = 0b100;
    else if (fabsf(odr - 30.0f) < 0.01f)   do_bits = 0b101;
    else if (fabsf(odr - 75.0f) < 0.01f)   do_bits = 0b110;
    else                                   return false;

    if (gain > 7) return false;

    uint8_t config_a = (ma << 5) | (do_bits << 2);
    _write_reg8(REG_CONFIG_A, config_a);

    uint8_t config_b = (gain << 5);
    _write_reg8(REG_CONFIG_B, config_b);

    _gain = gain;
    _gain_lsb_per_gauss = GAIN_LSB_PER_GAUSS[gain];
    return true;
}

bool HMC5883LFull::set_gain(uint8_t gain) {
    if (gain > 7) return false;
    _write_reg8(REG_CONFIG_B, gain << 5);
    _gain = gain;
    _gain_lsb_per_gauss = GAIN_LSB_PER_GAUSS[gain];
    return true;
}

bool HMC5883LFull::set_mode(const char* mode) {
    uint8_t md;
    if (strcmp(mode, "continuous") == 0)      md = 0b00;
    else if (strcmp(mode, "single") == 0)     md = 0b01;
    else if (strcmp(mode, "idle") == 0)       md = 0b10;
    else                                      return false;
    _write_reg8(REG_MODE, md);
    return true;
}

bool HMC5883LFull::data_ready() {
    uint8_t status = _read_reg8(REG_STATUS);
    return (status & 0x01) != 0;
}

uint8_t HMC5883LFull::status() {
    return _read_reg8(REG_STATUS);
}

bool HMC5883LFull::single_measurement(float& x, float& y, float& z) {
    _write_reg8(REG_MODE, 0x01);
    // Platform-specific 6 ms delay needed here
    return magnetic_field(x, y, z);
}

void HMC5883LFull::identify(uint8_t& id_a, uint8_t& id_b, uint8_t& id_c) {
    id_a = _read_reg8(REG_ID_A);
    id_b = _read_reg8(REG_ID_B);
    id_c = _read_reg8(REG_ID_C);
}

bool HMC5883LFull::self_test(bool positive, float& x, float& y, float& z) {
    uint8_t config_a = _read_reg8(REG_CONFIG_A);
    config_a = (config_a & 0xFC) | (positive ? 0b01 : 0b10);
    _write_reg8(REG_CONFIG_A, config_a);

    _write_reg8(REG_MODE, 0x01);
    // Platform-specific 6 ms delay needed here
    bool result = magnetic_field(x, y, z);

    config_a = (config_a & 0xFC) | 0b00;
    _write_reg8(REG_CONFIG_A, config_a);
    return result;
}